nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/serialization/CompressableRecordReader.java

```diff
@@ -0,0 +1,281 @@
+        this(in, filename, null, maxAttributeChars);
+    }
+
+    public CompressableRecordReader(final InputStream in, final String filename, final TocReader tocReader, final int maxAttributeChars) throws IOException {
+        logger.trace("Creating RecordReader for {}", filename);
+
+        rawInputStream = new ByteCountingInputStream(in);
+        this.maxAttributeChars = maxAttributeChars;
+
+        final InputStream limitedStream;
+        if (tocReader == null) {
+            limitedStream = rawInputStream;
+        } else {
+            final long offset1 = tocReader.getBlockOffset(1);
+            if (offset1 < 0) {
+                limitedStream = rawInputStream;
+            } else {
+                limitedStream = new LimitingInputStream(rawInputStream, offset1 - rawInputStream.getBytesConsumed());
+            }
+        }
+
+        final InputStream readableStream;
+        if (filename.endsWith(".gz")) {
+            readableStream = new BufferedInputStream(new GZIPInputStream(limitedStream));
+            compressed = true;
+        } else {
+            readableStream = new BufferedInputStream(limitedStream);
+            compressed = false;
+        }
+
+        byteCountingIn = new ByteCountingInputStream(readableStream);
+        dis = new DataInputStream(byteCountingIn);
+
+        final String repoClassName = dis.readUTF();
+        final int serializationVersion = dis.readInt();
+        headerLength = repoClassName.getBytes(StandardCharsets.UTF_8).length + 2 + 4; // 2 bytes for string length, 4 for integer.
+
+        if (serializationVersion < 1 || serializationVersion > 9) {
+            throw new IllegalArgumentException("Unable to deserialize record because the version is " + serializationVersion + " and supported versions are 1-9");
+        }
+
+        this.serializationVersion = serializationVersion;
+        this.filename = filename;
+        this.tocReader = tocReader;
+
+        readHeader(dis, serializationVersion);
+    }
+
+    @Override
+    public void skipToBlock(final int blockIndex) throws IOException {
+        if (tocReader == null) {
+            throw new IllegalStateException("Cannot skip to block " + blockIndex + " for Provenance Log " + filename + " because no Table-of-Contents file was found for this Log");
+        }
+
+        if (blockIndex < 0) {
+            throw new IllegalArgumentException("Cannot skip to block " + blockIndex + " because the value is negative");
+        }
+
+        if (blockIndex == getBlockIndex()) {
+            return;
+        }
+
+        final long offset = tocReader.getBlockOffset(blockIndex);
+        if (offset < 0) {
+            throw new IOException("Unable to find block " + blockIndex + " in Provenance Log " + filename);
+        }
+
+        final long curOffset = rawInputStream.getBytesConsumed();
+
+        final long bytesToSkip = offset - curOffset;
+        if (bytesToSkip >= 0) {
```

Comment:
No, we want to silently do nothing. This allows us to just call `skipToBlock(curBlock)` for subsequent records without having to add logic to determine if we are already on the proper block. I.e., this allows us to do something like `skipToBlock(1); readRecord(); skipToBlock(1); readRecord();` If we threw an Exception when bytesToSkip < 0 then the second call to `skipToBlock(1)` would fail since we would have already consumed data past the beginning of the block.